### PR TITLE
IAngleEncoder, EmptyAngleEncoder & CANCoderEncoder -> core

### PR DESCRIPTION
**Pull request by drorta**

### CODE REVIEW

File: src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java [hunk 1]
```diff
@@ -0,0 +1,10 @@
+package frc.robot.hardware.encoder;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+
+public class EmptyEncoder implements IEncoder {
```
Comment (bossyshrimp68) [inline src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java:5]: wdym EmptyEncoder?

**Review by bossyshrimp68 — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java (hunk 1 above)
Comment (YHerm) [reply src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java:5]: (its dror)if u like have a swerve sim or somth that gets an encoder but you dont have so ull use this instead (i dont fully understand this either)

**Review by GilStein1 — APPROVED** [review]
bro really thought I would approve that

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+
+	private final CANcoder encoder;
+
+	public CANCoderEncoder(CANcoder encoder) {
+		this.encoder = encoder;
+	}
+
+	@Override
+	public void setPosition(Rotation2d position) {
+		encoder.setPosition(position.getRotations());
```
Comment (hillelv) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:16]: format

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+	}
+
+}
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:19]: add connected inputs usage

File: src/main/java/frc/robot/hardware/encoder/IAngleEncoder.java [hunk 4]
```diff
@@ -0,0 +1,9 @@
+package frc.robot.hardware.encoder;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+
+public interface IAngleEncoder {
+
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/IAngleEncoder.java:6]: add isConnected()

**Review by YHerm — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java [hunk 5]
```diff
@@ -0,0 +1,25 @@
...
+	}
+
+	@Override
+	public void setPosition(Rotation2d position) {
+		encoder.setPosition(position.getRotations());
+	}
+
+	@Override
+	public boolean isConnected() {
+		return BaseStatusSignal.isAllGood(encoder.getPosition());
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:22]: Use connected inputAutoLogged

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+	}
+
+	@Override
+	public void updateInputs(ConnectedInputAutoLogged inputs){
+		inputs.connected = isConnected();
+	}
+
+	@Override
+	public void updateSignals(InputSignal... signals) {
+		for (InputSignal signal1: signals) {
```
Comment (hillelv) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:43]:
```suggestion
		for (InputSignal signal : signals) {
```

File: gradlew
Comment (hillelv) [inline gradlew:1]: check why its here

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+	public void updateInputs(ConnectedInputAutoLogged inputs) {
+		inputs.connected = isOK();
+	}
+
+	@Override
+	public void updateSignals(InputSignal... signals) {
+		for (InputSignal signal : signals) {
+			if (signal instanceof Phoenix6SignalBuilder.SignalGetter) {
+				BaseStatusSignal.refreshAll(((Phoenix6SignalBuilder.SignalGetter) signal).getSignal());
+			}
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:38]: u need special handle for BothLatency

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+
+	@Override
+	public boolean isOK() {
+		return BaseStatusSignal.isAllGood(encoder.getPosition());
+	}
+
+	@Override
+	public void updateInputs(ConnectedInputAutoLogged inputs) {
+		inputs.connected = isOK();
+	}
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:31]: remove this func

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java (hunk 7 above)
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:38]: i want u to save all signals and then refresh all at once

File: src/main/java/frc/robot/hardware/encoder/IAngleEncoder.java [hunk 9]
```diff
@@ -0,0 +1,17 @@
...
+import frc.robot.hardware.ConnectedInputAutoLogged;
+import frc.robot.hardware.signal.InputSignal;
+
+public interface IAngleEncoder {
+
+	void setPosition(Rotation2d position);
+
+	boolean isOK();
+
+	void updateInputs(ConnectedInputAutoLogged inputs);
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/IAngleEncoder.java:13]: no need

File: src/main/java/frc/robot/hardware/encoder/IAngleEncoder.java [hunk 10]
```diff
@@ -0,0 +1,17 @@
...
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import frc.robot.hardware.ConnectedInputAutoLogged;
+import frc.robot.hardware.signal.InputSignal;
+
+public interface IAngleEncoder {
+
+	void setPosition(Rotation2d position);
+
+	boolean isOK();
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/IAngleEncoder.java:11]:
```suggestion
	boolean isConnected();
```

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:1]: use the connected input inside this class instead of as param

**Review by YHerm — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java [hunk 11]
```diff
@@ -0,0 +1,54 @@
...
+import frc.robot.hardware.ConnectedInputAutoLogged;
+import frc.robot.hardware.IDevice;
+import frc.robot.hardware.signal.InputSignal;
+import frc.robot.hardware.signal.phoenix.Phoenix6BothLatencySignal;
+import frc.robot.hardware.signal.phoenix.Phoenix6SignalBuilder;
+import org.littletonrobotics.junction.Logger;
+
+import java.util.LinkedList;
+
+public class CANCoderEncoder implements IAngleEncoder, IDevice {
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:16]:
```suggestion
public class CANCoderEncoder extends Phoenix6Device implements IAngleEncoder, IDevice {
```

File: src/main/java/frc/robot/hardware/encoder/IAngleEncoder.java [hunk 12]
```diff
@@ -0,0 +1,9 @@
+package frc.robot.hardware.encoder;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+
+public interface IAngleEncoder {
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/IAngleEncoder.java:5]: maybe extend IDevice

Comment (drorta) [reply src/main/java/frc/robot/hardware/encoder/IAngleEncoder.java:5]: What will it do?

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+import edu.wpi.first.math.geometry.Rotation2d;
+import frc.robot.hardware.ConnectedInputAutoLogged;
+import frc.robot.hardware.Phoenix6Device;
+import frc.robot.hardware.signal.InputSignal;
+
+
+public class CANCoderEncoder extends Phoenix6Device implements IAngleEncoder {
+
+	private final CANcoder encoder;
+	private final ConnectedInputAutoLogged connectedInput;
```
Comment (greenblitz4590) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:13]: remove

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java [hunk 14]
```diff
@@ -0,0 +1,38 @@
...
+	private final String logPath;
+
+	public CANCoderEncoder(CANcoder encoder, String logPath) {
+		super(logPath);
+		this.encoder = encoder;
+		this.connectedInput = new ConnectedInputAutoLogged();
```
Comment (greenblitz4590) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:19]: also remove

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java [hunk 15]
```diff
@@ -0,0 +1,38 @@
...
+import frc.robot.hardware.ConnectedInputAutoLogged;
+import frc.robot.hardware.Phoenix6Device;
+import frc.robot.hardware.signal.InputSignal;
+
+
+public class CANCoderEncoder extends Phoenix6Device implements IAngleEncoder {
+
+	private final CANcoder encoder;
+	private final ConnectedInputAutoLogged connectedInput;
+	private final String logPath;
```
Comment (greenblitz4590) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:14]: also remove

**Review by greenblitz4590 — CHANGES_REQUESTED** [review]

**Review by hillelv — APPROVED** [review]

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+
+	@Override
+	public boolean isConnected() {
+		return super.isConnected();
+	}
+
+	@Override
+	public void updateSignals(InputSignal... signals) {
+		super.updateSignals(signals);
+	}
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:30]: this is completely unneeded

File: src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java [hunk 17]
```diff
@@ -0,0 +1,19 @@
+package frc.robot.hardware.encoder;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import frc.robot.hardware.signal.InputSignal;
+
+public class EmptyEncoder implements IAngleEncoder {
```
Comment (YoniKiriaty) [inline src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java:6]: consider changing the name to `EmptyAngleEncoder`

File: src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java
Comment (YoniKiriaty) [inline src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java:1]: what is the use for this object?

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java (hunk 16 above)
Comment (drorta) [reply src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:30]: what?

File: src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java
Comment (drorta) [reply src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java:1]: Yoav said replay or smth, I don't really understand myself

**Review by YHerm — CHANGES_REQUESTED** [review]
@drorta

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java (hunk 16 above)
Comment (drorta) [reply src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:30]: oh lol

File: src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java
Comment (drorta) [reply src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java:1]: @YoniKiriaty

File: src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java (hunk 17 above)
[the remark this answers is not in the document]
Comment (drorta) [reply src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java:6]: whats the branch i need to merge with to have emptydevice?

[the remark this answers is not in the document]
Comment (drorta) [reply src/main/java/frc/robot/hardware/encoder/EmptyEncoder.java:6]: ctre device?

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+import frc.robot.hardware.Phoenix6Device;
+import frc.robot.hardware.signal.InputSignal;
+
+public class CANCoderEncoder extends Phoenix6Device implements IAngleEncoder {
+
+	private final CANcoder encoder;
+
+	public CANCoderEncoder(CANcoder encoder, String logPath) {
+		super(logPath);
+		this.encoder = encoder;
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:14]: add optimize bus util

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+import com.ctre.phoenix6.hardware.CANcoder;
+import edu.wpi.first.math.geometry.Rotation2d;
+import frc.robot.hardware.Phoenix6Device;
+import frc.robot.hardware.signal.InputSignal;
+
+public class CANCoderEncoder extends Phoenix6Device implements IAngleEncoder {
+
+	private final CANcoder encoder;
+
+	public CANCoderEncoder(CANcoder encoder, String logPath) {
```
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:12]: log path first

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java
Comment (YHerm) [inline src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java:1]: rename package angle encoder

**Review by YHerm — CHANGES_REQUESTED** [review]

File: src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java (hunk 18 above)
Comment (drorta) [reply src/main/java/frc/robot/hardware/encoder/CANCoderEncoder.java]: do i need to test again?

**Review by YHerm — APPROVED** [review]
Make sure u test all

**Review by NoamRosenberg08 — APPROVED** [review]

**Review by bossyshrimp68 — APPROVED** [review]

**Review by YHerm — APPROVED** [review]

File: src/main/java/frc/robot/hardware/angleencoder/EmptyAngleEncoder.java [hunk 20]
```diff
@@ -0,0 +1,15 @@
+package frc.robot.hardware.angleencoder;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import frc.robot.hardware.EmptyDevice;
+
+public class EmptyAngleEncoder extends EmptyDevice implements IAngleEncoder {
+
+	public EmptyAngleEncoder(String logPath) {
```
Comment (GilStein1) [inline src/main/java/frc/robot/hardware/angleencoder/EmptyAngleEncoder.java:8]: maybe EmptyEncoder? the angle word is מובן מאליו

File: src/main/java/frc/robot/hardware/angleencoder/IAngleEncoder.java [hunk 21]
```diff
@@ -0,0 +1,10 @@
+package frc.robot.hardware.angleencoder;
+
+import edu.wpi.first.math.geometry.Rotation2d;
+import frc.robot.hardware.IDevice;
+
+public interface IAngleEncoder extends IDevice {
```
Comment (GilStein1) [inline src/main/java/frc/robot/hardware/angleencoder/IAngleEncoder.java:6]: IEncoder

**Review by bossyshrimp68 — APPROVED** [review]

File: src/main/java/frc/robot/hardware/angleencoder/CANCoderEncoder.java [hunk 22]
```diff
@@ -0,0 +1,22 @@
...
+import frc.robot.hardware.phoenix6.Phoenix6Device;
+
+public class CANCoderEncoder extends Phoenix6Device implements IAngleEncoder {
+
+	private final CANcoder encoder;
+
+	public CANCoderEncoder(String logPath, CANcoder encoder) {
+		super(logPath);
+		this.encoder = encoder;
+		encoder.optimizeBusUtilization();
```
Comment (hillelv) [inline src/main/java/frc/robot/hardware/angleencoder/CANCoderEncoder.java:14]:
```suggestion
		this.encoder.optimizeBusUtilization();
```
לא חייב

**Review by hillelv — APPROVED** [review]

File: src/main/java/frc/robot/hardware/angleencoder/CANCoderEncoder.java (hunk 22 above)
Comment (YHerm) [reply src/main/java/frc/robot/hardware/angleencoder/CANCoderEncoder.java:14]: no need this when using, only when setting the member

File: src/main/java/frc/robot/hardware/angleencoder/EmptyAngleEncoder.java (hunk 20 above)
Comment (GilStein1) [reply src/main/java/frc/robot/hardware/angleencoder/EmptyAngleEncoder.java:8]: bruh

Comment (GilStein1) [reply src/main/java/frc/robot/hardware/angleencoder/EmptyAngleEncoder.java:8]: idk if distance encoder is a correct name

**Review by NoamRosenberg08 — APPROVED** [review]

**Pull request merged**